### ISSUE TITLE
[java] UnusedAssignment false positive in record compact constructor

### ISSUE THREAD

**Issue by kdebski85** [pmd/pmd#4603]
**Affects PMD Version:** 7.0.0-rc3 (works fine in 6.55.0)

**Rule:** [UnusedAssignment](https://docs.pmd-code.org/latest/pmd_rules_java_bestpractices.html#unusedassignment)

**Description:**

False-positive is reported when a variable is assigned in record compact constructor.

```
 UnusedAssignment:       The value assigned to variable 'foo' is never used
```

**Code Sample demonstrating the issue:**

```java
class Main {
    public static void main(String[] args) {
        System.out.println(new TestRecord(2).foo());
    }
}

record TestRecord(int foo) {
    TestRecord {
        foo = Math.min(foo, 1); // <!--- violation here
    }
}
```

**Expected outcome:**

PMD should not report false-positive about unused `foo` in compact constructor.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*
Gradle
